File: templates/problems/07_trees/07_lca_bst/go/lca_bst_test.go
package lca_bst

import "testing"

func TestLowestCommonAncestor(t *testing.T) {
	tests := []struct {
		name   string
		vals   []int
		pVal   int
		qVal   int
		wantV  int
	}{
		{"example 1", []int{6, 2, 8, 0, 4, 7, 9, -101, -101, 3, 5}, 2, 8, 6},
		{"example 2", []int{6, 2, 8, 0, 4, 7, 9, -101, -101, 3, 5}, 2, 4, 2},
		{"root is lca", []int{2, 1, 3}, 1, 3, 2},
		{"same node", []int{2, 1, 3}, 1, 1, 1},
		{"right subtree", []int{6, 2, 8, 0, 4, 7, 9}, 7, 9, 8},
		{"deep nodes", []int{6, 2, 8, 0, 4, 7, 9, -101, -101, 3, 5}, 3, 5, 4},
	}

	// findNode finds a node by value in the tree.
	var findNode func(root *TreeNode, val int) *TreeNode
	findNode = func(root *TreeNode, val int) *TreeNode {
		if root == nil || root.Val == val {
			return root
		}
		if l := findNode(root.Left, val); l != nil {
			return l
		}
		return findNode(root.Right, val)
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			root := buildTreeHelper(tt.vals)
			p := findNode(root, tt.pVal)
			q := findNode(root, tt.qVal)
			got := lowestCommonAncestor(root, p, q)
			if got == nil || got.Val != tt.wantV {
				gotV := -1
				if got != nil {
					gotV = got.Val
				}
				t.Errorf("lowestCommonAncestor() = %v, want %v", gotV, tt.wantV)
			}
		})
	}
}

File: templates/problems/07_trees/07_lca_bst/typescript/lca_bst.test.ts
import { describe, it, expect } from "vitest";
import { buildTree, TreeNode } from "./helpers";
import { lowestCommonAncestor } from "./lca_bst";

function findNode(root: TreeNode | null, val: number): TreeNode | null {
    if (root === null || root.val === val) {
        return root;
    }
    const left = findNode(root.left, val);
    if (left !== null) {
        return left;
    }
    return findNode(root.right, val);
}

describe("lowestCommonAncestor", () => {
    it.each([
        { name: "example 1", vals: [6, 2, 8, 0, 4, 7, 9, -101, -101, 3, 5], pVal: 2, qVal: 8, wantV: 6 },
        { name: "example 2", vals: [6, 2, 8, 0, 4, 7, 9, -101, -101, 3, 5], pVal: 2, qVal: 4, wantV: 2 },
        { name: "root is lca", vals: [2, 1, 3], pVal: 1, qVal: 3, wantV: 2 },
        { name: "same node", vals: [2, 1, 3], pVal: 1, qVal: 1, wantV: 1 },
        { name: "right subtree", vals: [6, 2, 8, 0, 4, 7, 9], pVal: 7, qVal: 9, wantV: 8 },
        { name: "deep nodes", vals: [6, 2, 8, 0, 4, 7, 9, -101, -101, 3, 5], pVal: 3, qVal: 5, wantV: 4 },
    ])("case: $name", ({ vals, pVal, qVal, wantV }) => {
        const root = buildTree(vals);
        const p = findNode(root, pVal)!;
        const q = findNode(root, qVal)!;
        const got = lowestCommonAncestor(root, p, q);
        const gotV = got === null ? -1 : got.val;
        expect(gotV).toBe(wantV);
    });
});
